Adds Unit Test for URDF Parser's ParseMaterial() Method.

liangfok opened the pull request:
Having more unit tests is desirable for software maintainability.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/4222)
<!-- Reviewable:end -->

jwnimmer-tri commented:
Dup of #4221, I think.  (And this has some weird merge artifacts?)

jwnimmer-tri commented:
Though we have spelled the `lpNorm` differently.  I hope that my spellings are okay?  (I spot-checked a few of them with manual testing.)

liangfok commented:
Looks like there is some overlap. I'll let yours go in first and then update this PR to include the updated unit test.

liangfok commented:
@jwnimmer-tri I think your way of computing the infinity norm is correct. I accidentally used the "operator infinity norm".

liangfok commented:
I fixed the wierd git merge effects. Now waiting on #4221 to merge before submitting this PR for review.

---

Review status: 0 of 7 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-:-KXCYMmVH9LcHdCMLfTr:b-d294h7)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Yeah, sorry, Jenkins is slow on #4221.  (It missed some triggers, so I had to re-launch them by hand.)  Hopefully its ready to merge within half an hour or so.

jwnimmer-tri commented:
-@jwnimmer-tri feature.  I'm probably not the right person to feature-review multibody parser test additions?

Also, please update the OP summary of the PR now that its been rebased.

---

Review status: 0 of 6 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-:-KXCpu3zJg5UJytaMq_B:b-mpuq4j)*
<!-- Sent from Reviewable.io -->

liangfok commented:
+@amcastro-tri for feature review please.

---

Review status: 0 of 6 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-:-KXCrkA-PWFfWUGIkRxW:b-57wv7n)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Review status: 0 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/multibody/test/CMakeLists.txt, line 56 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-KXCz52PscRkojfiz_PC:-KXCz52QT27YZpACS5ba:bmh212z) ([raw file](https://github.com/robotlocomotion/drake/blob/f87cb3f309e9a80f82e7830be69f6516219e6a23/drake/multibody/test/CMakeLists.txt#L56)):*
> ```CMake
> endif ()
> 
> add_subdirectory(parser_urdf_test)
> ```

BTW, why the change to Yoda-order?

---

*[drake/multibody/test/parser_urdf_test/conflicting_materials.urdf, line 1 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-KXCzH3ybt7buxH5vSWR:-KXCzH3ybt7buxH5vSWS:b5mhpkz) ([raw file](https://github.com/robotlocomotion/drake/blob/f87cb3f309e9a80f82e7830be69f6516219e6a23/drake/multibody/test/parser_urdf_test/conflicting_materials.urdf#L1)):*
> ```XML
> <robot name="non_conflicting_materials">
> ```

did you mean `conflicting`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4222)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Reviewed 7 of 7 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4222)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/multibody/test/CMakeLists.txt, line 56 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-KXCz52PscRkojfiz_PC:-KXD-DeEL43SGGTYmSDx:bthgfe8) ([raw file](https://github.com/robotlocomotion/drake/blob/f87cb3f309e9a80f82e7830be69f6516219e6a23/drake/multibody/test/CMakeLists.txt#L56)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> BTW, why the change to Yoda-order?
</details>
FYI That's what the DUT file is named (`parser_urdf.h`).  Sigh.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4222)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 5 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/multibody/test/CMakeLists.txt, line 56 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-KXCz52PscRkojfiz_PC:-KXDtwTSVx_ZyqZJK_AT:b5r5tn) ([raw file](https://github.com/robotlocomotion/drake/blob/f87cb3f309e9a80f82e7830be69f6516219e6a23/drake/multibody/test/CMakeLists.txt#L56)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> FYI That's what the DUT file is named (`parser_urdf.h`).  Sigh.
</details>
Let's punt the re-name of the parsers to a future PR. Perhaps along with the resolution of #4091.

---

*[drake/multibody/test/parser_urdf_test/conflicting_materials.urdf, line 1 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-KXCzH3ybt7buxH5vSWR:-KXDtuuir0x9E6LQqV2d:boyexbg) ([raw file](https://github.com/robotlocomotion/drake/blob/f87cb3f309e9a80f82e7830be69f6516219e6a23/drake/multibody/test/parser_urdf_test/conflicting_materials.urdf#L1)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> did you mean `conflicting`?
</details>
Done. Thanks.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4222)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Just two small comments

---

Reviewed 5 of 7 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/multibody/test/parser_urdf_test/non_conflicting_materials_2.urdf, line 4 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-KXfLDp1ld1PRukxReFv:-KXfLDp2uLkwLLELD4-4:bms7g5s) ([raw file](https://github.com/robotlocomotion/drake/blob/5006498aba723a8e1e9655f40acb69bcd91364a9/drake/multibody/test/parser_urdf_test/non_conflicting_materials_2.urdf#L4)):*
> ```XML
>   <material name="brown">
>     <color rgba="0.93333333333 0.79607843137 0.67843137254 1"/>
>   </material>
> ```

BTW, So..... having exactly the same entry multiple times is not a defect? is that part of the URDF specification? I think better would be the parser to throw an error if there are two materials defined with the same name even if just copy/pasted.

---

*[drake/multibody/test/parser_urdf_test/parser_urdf_test.cc, line 97 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-KXfM9IaU-WCO-JTQjur:-KXfM9IaU-WCO-JTQjus:b-mfasxe) ([raw file](https://github.com/robotlocomotion/drake/blob/5006498aba723a8e1e9655f40acb69bcd91364a9/drake/multibody/test/parser_urdf_test/parser_urdf_test.cc#L97)):*
> ```cpp
>       file_no_conflict_1, tree.get()));
> 
>   tree.reset(new RigidBodyTree<double>());
> ```

Doing `make_unique<RigidBodyTree<double>>()` here (and bellow) doesn't have the same effect as reset? (and we'd avoid the bare `new` for consistency with the rest of our codebase).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-:-KXfMPsLWzWSr5_DNd3P:b7izv3k)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/multibody/test/parser_urdf_test/non_conflicting_materials_2.urdf, line 4 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-KXfLDp1ld1PRukxReFv:-KXfb2P6fXK2H94bKwwB:b-t4rdd4) ([raw file](https://github.com/robotlocomotion/drake/blob/5006498aba723a8e1e9655f40acb69bcd91364a9/drake/multibody/test/parser_urdf_test/non_conflicting_materials_2.urdf#L4)):*
> URDF specification

:laughing: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-:-KXfb8FLxu_gjo0UmEXz:bnfp4nl)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 2 of 8 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/multibody/test/parser_urdf_test/non_conflicting_materials_2.urdf, line 4 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-KXfLDp1ld1PRukxReFv:-KXfNtR5mZU1s1qgqBbm:b-mhriwf) ([raw file](https://github.com/robotlocomotion/drake/blob/5006498aba723a8e1e9655f40acb69bcd91364a9/drake/multibody/test/parser_urdf_test/non_conflicting_materials_2.urdf#L4)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

> URDF specification

:laughing: 
</blockquote></details>

The [official URDF website](http://wiki.ros.org/urdf/XML/link) is vague regarding the proper behavior when the parser encounters multiple identical material specifications. Out of curiosity, I loaded this model into Gazebo and confirmed that it aborts. I've thus updated this PR to match Gazebo's behavior.

Using Gazebo, I also learned that materials defined within a `link` _can_ clash with materials defined within a `robot` and have thus modified the URDF parser to reflect this behavior.

---

*[drake/multibody/test/parser_urdf_test/parser_urdf_test.cc, line 97 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-KXfM9IaU-WCO-JTQjur:-KXfrlNSJV5-fRGpL-56:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5006498aba723a8e1e9655f40acb69bcd91364a9/drake/multibody/test/parser_urdf_test/parser_urdf_test.cc#L97)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Doing `make_unique<RigidBodyTree<double>>()` here (and bellow) doesn't have the same effect as reset? (and we'd avoid the bare `new` for consistency with the rest of our codebase).
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4222)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I addressed all reviewer comments. PTAL. Thanks.

---

Review status: 2 of 8 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-:-KXgEiGiRYhfP-5asi9u:bblnx7v)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
+(status: curate commits before merging) 

---

Review status: 2 of 8 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/multibody/parser_urdf.cc, line 389 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-KXgH8vDn42bvEAfHYmO:-KXgH8vDn42bvEAfHYmP:b19vzs8) ([raw file](https://github.com/robotlocomotion/drake/blob/a6bd65882df20de20d5bed49fc62c7ca08b88bff/drake/multibody/parser_urdf.cc#L389)):*
> ```cpp
>     // (for example, see this URDF by Robotiq: http://bit.ly/28P0pmo).
>     if (color_specified && name_specified)
>       AddMaterialToMaterialMap(material_name, rgba, false, materials);
> ```

Is this the only place you set `throw_if_name_clash` to false?  Does anything actually break if you set it to `true`?  Since you've established that `true` aligns with Gazebo, I'm hoping we can just get rid of the option altogether.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-:-KXgGdJcgJfXcf_MzVHK:b-s794vx)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I addressed all reviewer comments. PTAL. Thanks.

---

Review status: 2 of 8 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/multibody/parser_urdf.cc, line 389 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-KXgH8vDn42bvEAfHYmO:-KXgL1bHfgnrFGaaAiPJ:b-ibiv4v) ([raw file](https://github.com/robotlocomotion/drake/blob/a6bd65882df20de20d5bed49fc62c7ca08b88bff/drake/multibody/parser_urdf.cc#L389)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

Is this the only place you set `throw_if_name_clash` to false?  Does anything actually break if you set it to `true`?  Since you've established that `true` aligns with Gazebo, I'm hoping we can just get rid of the option altogether.
</blockquote></details>

Unfortunately, if we set the `throw_if_name_clash` parameter to be `true`, Robotiq's URDF (mentioned in the comment preceding this code), will not load because it defines the same color multiple times in different links. I updated the unit test to cover this URDF.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-:-KXgWY88qAECAQ9jC7Sa:bblnx7v)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 2 of 8 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/multibody/parser_urdf.cc, line 389 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-KXgH8vDn42bvEAfHYmO:-KXg_g3A5Q_D0wpMR6HR:b-nosw13) ([raw file](https://github.com/robotlocomotion/drake/blob/a6bd65882df20de20d5bed49fc62c7ca08b88bff/drake/multibody/parser_urdf.cc#L389)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Unfortunately, if we set the `throw_if_name_clash` parameter to be `true`, Robotiq's URDF (mentioned in the comment preceding this code), will not load because it defines the same color multiple times in different links. I updated the unit test to cover this URDF.
</blockquote></details>

I would agree with @david-german-tri here. Is the file you mention above generated properly? it says it was automatically generated from a xacro file. Is that a problem with xacro generated files in general? if that is the case then this comment should probably reflect that rather than a random urdf file from the web. 
I mean, is this a problem with the parser (and then I agree it needs fixing) or with the urdf you are trying to read?

---

*[drake/multibody/parser_urdf.cc, line 131 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-KXgZlyv45_G8mryDa5d:-KXgZlywTnGFcJFwt_aO:btowkg5) ([raw file](https://github.com/robotlocomotion/drake/blob/c5df637048624de176745ad5b942442f53d7fac9/drake/multibody/parser_urdf.cc#L131)):*
> ```cpp
>                  << "  - new RGBA values: " << color_rgba.transpose()
>                  << std::endl;
>       throw std::runtime_error(error_buff.str());
> ```

I think this is a good place for `DRAKE_ABORT_MSG` to have consistent error messages.

---

*[drake/multibody/parser_urdf.cc, line 139 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-KXgakI7nStHlKlYjiIj:-KXgakI8aHMD24c0obb8:b-jrnjjx) ([raw file](https://github.com/robotlocomotion/drake/blob/c5df637048624de176745ad5b942442f53d7fac9/drake/multibody/parser_urdf.cc#L139)):*
> ```cpp
> }
> 
> void ParseMaterial(XMLElement* node, bool throw_if_name_clash,
> ```

Maybe I can live with the `throw_if_clash` parameter for `AddMaterialToMaterialMap` if it solves a problem for xacro generated files but, why does `ParseMaterial` need it as an argument? where in this PR is `ParseMaterial` called with `throw_if_clash=false`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4222)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Review status: 2 of 8 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/multibody/parser_urdf.cc, line 139 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-KXgYZbier3ZmyG416VG:-KXgYZbier3ZmyG416VH:bvwfy52) ([raw file](https://github.com/robotlocomotion/drake/blob/c5df637048624de176745ad5b942442f53d7fac9/drake/multibody/parser_urdf.cc#L139)):*
> ```cpp
> }
> 
> void ParseMaterial(XMLElement* node, bool throw_if_name_clash,
> ```

So, let me see if I've got this.  We know there are nonconforming URDFs (such as `robotiq.urdf`) that declare the same material in multiple `visual` elements.   Those URDFs crash Gazebo, but we want to parse them anyway.  Therefore, we have this `throw_if_name_clash` knob in `AddMaterialToMaterialMap`.  

We do not expose it as an argument to `ParseVisual`; instead, we always set it to false.  We do expose it as an argument to `ParseMaterial`, but at all existing call sites the value of that argument is true.  This configuration works for all the URDFs we care about right now, because the only ones that duplicate materials do so in `visual` elements.  Is that right?

If so, at a minimum I think we should get rid of this argument as @amcastro-tri suggests below.  Going a step farther, should we consider being permissive of name clashes in all situations?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4222)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I addressed all reviewer comments. PTAL. Thanks.

---

Review status: 2 of 8 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/multibody/parser_urdf.cc, line 389 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-KXgH8vDn42bvEAfHYmO:-KXgeM0-yr57t9WtHcIy:b-ngya6g) ([raw file](https://github.com/robotlocomotion/drake/blob/a6bd65882df20de20d5bed49fc62c7ca08b88bff/drake/multibody/parser_urdf.cc#L389)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

I would agree with @david-german-tri here. Is the file you mention above generated properly? it says it was automatically generated from a xacro file. Is that a problem with xacro generated files in general? if that is the case then this comment should probably reflect that rather than a random urdf file from the web. 
I mean, is this a problem with the parser (and then I agree it needs fixing) or with the urdf you are trying to read?

</blockquote></details>

Xacro can pull in multiple other `.xacro` files when producing a URDF one so I can see how duplicate material specifications may show up in a single URDF. Based on more recent testing, I found that Gazebo supports duplicate material specifications as long as they are not all at the `<robot>` level (see comment in the other thread).

---

*[drake/multibody/parser_urdf.cc, line 131 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-KXgZlyv45_G8mryDa5d:-KXgi4-1jscTbB-HtICQ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/c5df637048624de176745ad5b942442f53d7fac9/drake/multibody/parser_urdf.cc#L131)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

I think this is a good place for `DRAKE_ABORT_MSG` to have consistent error messages.
</blockquote></details>

Done.

---

*[drake/multibody/parser_urdf.cc, line 139 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-KXgYZbier3ZmyG416VG:-KXgdUouEm47sleFlQrH:b-gukdid) ([raw file](https://github.com/robotlocomotion/drake/blob/c5df637048624de176745ad5b942442f53d7fac9/drake/multibody/parser_urdf.cc#L139)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

So, let me see if I've got this.  We know there are nonconforming URDFs (such as `robotiq.urdf`) that declare the same material in multiple `visual` elements.   Those URDFs crash Gazebo, but we want to parse them anyway.  Therefore, we have this `throw_if_name_clash` knob in `AddMaterialToMaterialMap`.  

We do not expose it as an argument to `ParseVisual`; instead, we always set it to false.  We do expose it as an argument to `ParseMaterial`, but at all existing call sites the value of that argument is true.  This configuration works for all the URDFs we care about right now, because the only ones that duplicate materials do so in `visual` elements.  Is that right?

If so, at a minimum I think we should get rid of this argument as @amcastro-tri suggests below.  Going a step farther, should we consider being permissive of name clashes in all situations?
</blockquote></details>

It turns out that `drake/examples/Atlas/urdf/robotiq.urdf` is conformant in terms of being loadable in Gazebo (see screenshot below):

![Screenshot from 2016-11-28 15:46:23.png](https://files.reviewableusercontent.io/images/reviewable/1388098/FutL1-3VnPIbN_UATxv531A5/Screenshot%20from%202016-11-28%2015%3A46%3A23.png)

Apparently, in this case, Gazebo does not crash because the duplicate material specifications are in different links (they are not directly under the `robot` tag, which would cause Gazebo to crash).

---

*[drake/multibody/parser_urdf.cc, line 139 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-KXgakI7nStHlKlYjiIj:-KXgeDUaSDeTWpf9Tunq:bhvtz04) ([raw file](https://github.com/robotlocomotion/drake/blob/c5df637048624de176745ad5b942442f53d7fac9/drake/multibody/parser_urdf.cc#L139)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Maybe I can live with the `throw_if_clash` parameter for `AddMaterialToMaterialMap` if it solves a problem for xacro generated files but, why does `ParseMaterial` need it as an argument? where in this PR is `ParseMaterial` called with `throw_if_clash=false`?
</blockquote></details>

The `false` scenario is when parsing the `visual` elements within `link` elements. Based on testing with Gazebo, duplicate material specifications are allowed as long as they are in different links and not at the robot level.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-:-KXgkSSX-MjQu6Twh3BA:bblnx7v)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 2 of 8 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/multibody/parser_urdf.cc, line 139 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-KXgYZbier3ZmyG416VG:-KXgm8tXhnWZCkVCTuSA:b-kx52nl) ([raw file](https://github.com/robotlocomotion/drake/blob/c5df637048624de176745ad5b942442f53d7fac9/drake/multibody/parser_urdf.cc#L139)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

It turns out that `drake/examples/Atlas/urdf/robotiq.urdf` is conformant in terms of being loadable in Gazebo (see screenshot below):

![Screenshot from 2016-11-28 15:46:23.png](https://files.reviewableusercontent.io/images/reviewable/1388098/FutL1-3VnPIbN_UATxv531A5/Screenshot%20from%202016-11-28%2015%3A46%3A23.png)

Apparently, in this case, Gazebo does not crash because the duplicate material specifications are in different links (they are not directly under the `robot` tag, which would cause Gazebo to crash).
</blockquote></details>

I removed `abort_if_name_clash` from `ParseMaterial()`. Thanks!

---

*[drake/multibody/parser_urdf.cc, line 139 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-KXgakI7nStHlKlYjiIj:-KXgmDleNm4FkLaPDvN1:b-kx52nl) ([raw file](https://github.com/robotlocomotion/drake/blob/c5df637048624de176745ad5b942442f53d7fac9/drake/multibody/parser_urdf.cc#L139)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

The `false` scenario is when parsing the `visual` elements within `link` elements. Based on testing with Gazebo, duplicate material specifications are allowed as long as they are in different links and not at the robot level.
</blockquote></details>

I removed `abort_if_name_clash` from `ParseMaterial()`. Thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4222)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 2 of 8 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/multibody/parser_urdf.cc, line 389 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-KXgH8vDn42bvEAfHYmO:-KXgla9OYz1xTTFRkq3_:bi84sis) ([raw file](https://github.com/robotlocomotion/drake/blob/a6bd65882df20de20d5bed49fc62c7ca08b88bff/drake/multibody/parser_urdf.cc#L389)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Xacro can pull in multiple other `.xacro` files when producing a URDF one so I can see how duplicate material specifications may show up in a single URDF. Based on more recent testing, I found that Gazebo supports duplicate material specifications as long as they are not all at the `<robot>` level (see comment in the other thread).
</blockquote></details>

I see. Could you add that observation to this comment so we don't loose it? So it doen't seem to be a urdf specification issue (very loose by the way) but a xacro generation issue and I think that should be documented here.

---

*[drake/multibody/parser_urdf.cc, line 139 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-KXgakI7nStHlKlYjiIj:-KXgkyZ4YOKGJy7mzhmm:b-mncrqa) ([raw file](https://github.com/robotlocomotion/drake/blob/c5df637048624de176745ad5b942442f53d7fac9/drake/multibody/parser_urdf.cc#L139)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

The `false` scenario is when parsing the `visual` elements within `link` elements. Based on testing with Gazebo, duplicate material specifications are allowed as long as they are in different links and not at the robot level.
</blockquote></details>

yes, but I don't think `ParseMaterial` is being called with `false` anywhere in this PR?

---

*[drake/multibody/parser_urdf.cc, line 139 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-KXgYZbier3ZmyG416VG:-KXgm1mEOme9XNsTqArA:besvfuk) ([raw file](https://github.com/robotlocomotion/drake/blob/c5df637048624de176745ad5b942442f53d7fac9/drake/multibody/parser_urdf.cc#L139)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

I removed `abort_if_name_clash` from `ParseMaterial()`. Thanks!
</blockquote></details>

@liangfok, as @david-german-tri pointed out, my comment is not about removing the flag you added (which clearly makes these files loadable) but just not to have it as an argument in `ParseMaterial` since at all sites the flag is set to `true`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4222)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Ah! we both replied at the same time. Thanks for fixing. <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Review status: 2 of 8 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-:-KXgmuG9bwZ5oT7wFwt6:b-7c9lbx)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I addressed all reviewer comments. PTAL. Thanks.

---

Review status: 2 of 8 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/multibody/parser_urdf.cc, line 389 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-KXgH8vDn42bvEAfHYmO:-KXgnDDxKmB5XGBpdw7k:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a6bd65882df20de20d5bed49fc62c7ca08b88bff/drake/multibody/parser_urdf.cc#L389)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

I see. Could you add that observation to this comment so we don't loose it? So it doen't seem to be a urdf specification issue (very loose by the way) but a xacro generation issue and I think that should be documented here.
</blockquote></details>

Done.

---

*[drake/multibody/parser_urdf.cc, line 139 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-KXgYZbier3ZmyG416VG:-KXgnHCyOMrPZ5w_E9Sf:boyexbg) ([raw file](https://github.com/robotlocomotion/drake/blob/c5df637048624de176745ad5b942442f53d7fac9/drake/multibody/parser_urdf.cc#L139)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

@liangfok, as @david-german-tri pointed out, my comment is not about removing the flag you added (which clearly makes these files loadable) but just not to have it as an argument in `ParseMaterial` since at all sites the flag is set to `true`.
</blockquote></details>

Done. Thanks.

---

*[drake/multibody/parser_urdf.cc, line 139 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-KXgakI7nStHlKlYjiIj:-KXgnI-KTALgu0OdRYLT:boyexbg) ([raw file](https://github.com/robotlocomotion/drake/blob/c5df637048624de176745ad5b942442f53d7fac9/drake/multibody/parser_urdf.cc#L139)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

yes, but I don't think `ParseMaterial` is being called with `false` anywhere in this PR?
</blockquote></details>

Done. Thanks.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-:-KXgnKIQ0_XMAir4z_-m:bblnx7v)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Reviewed 1 of 7 files at r1, 4 of 6 files at r3, 1 of 2 files at r6.
Review status: 7 of 8 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/multibody/test/parser_urdf_test/non_conflicting_materials_2.urdf, line 6 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-KXh1FpYIwRmDdNuRSaB:-KXh1FpYIwRmDdNuRSaC:b-wn5kzw) ([raw file](https://github.com/robotlocomotion/drake/blob/a912be50fabc3bee748dbf49d9b0ba787ddf7b42/drake/multibody/test/parser_urdf_test/non_conflicting_materials_2.urdf#L6)):*
> ```XML
> This URDF contains a robot-level material named "brown" that clashes with a
> link-level material named "brown". When loading this, no exception should be
> thrown since the RGBA values for both are the same.
> ```

If you changed the order, so that the `link` material preceded the `robot` material, would this test still pass?

---

*[drake/multibody/test/parser_urdf_test/parser_urdf_test.cc, line 113 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-KXh1dMq-8WLYRdWGPqa:-KXh1dMq-8WLYRdWGPqb:bgnkznb) ([raw file](https://github.com/robotlocomotion/drake/blob/a912be50fabc3bee748dbf49d9b0ba787ddf7b42/drake/multibody/test/parser_urdf_test/parser_urdf_test.cc#L113)):*
> ```cpp
>   // This URDF defines the same color multiple times in different links.
>   const std::string file_robotiq = drake::GetDrakePath() +
>       "/examples/Atlas/urdf/robotiq.urdf";
> ```

This is a layering violation; instead, please make a test example that accurately reproduces what is interesting about the `robotiq.urdf`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4222)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I addressed all reviewer comments. PTAL. Thanks.

---

Review status: 6 of 10 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/multibody/test/parser_urdf_test/non_conflicting_materials_2.urdf, line 6 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-KXh1FpYIwRmDdNuRSaB:-KXhBlSFVg0MhkNFU1sd:b-oitnha) ([raw file](https://github.com/robotlocomotion/drake/blob/a912be50fabc3bee748dbf49d9b0ba787ddf7b42/drake/multibody/test/parser_urdf_test/non_conflicting_materials_2.urdf#L6)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

If you changed the order, so that the `link` material preceded the `robot` material, would this test still pass?
</blockquote></details>

Yes because the `<robot>` level materials are always parsed first before any of the `<link>` elements. I added an example URDF and test coverage for this scenario.

---

*[drake/multibody/test/parser_urdf_test/parser_urdf_test.cc, line 113 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-KXh1dMq-8WLYRdWGPqa:-KXhgYNMt5L3atF3vblo:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a912be50fabc3bee748dbf49d9b0ba787ddf7b42/drake/multibody/test/parser_urdf_test/parser_urdf_test.cc#L113)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

This is a layering violation; instead, please make a test example that accurately reproduces what is interesting about the `robotiq.urdf`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-:-KXhgb6RNW2avB3fAcXT:bblnx7v)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 6 of 10 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/multibody/test/parser_urdf_test/duplicate_but_same_materials.urdf, line 19 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-KXhh4ytYUctDxOEIykC:-KXhh4ytYUctDxOEIykD:b-fwinqq) ([raw file](https://github.com/robotlocomotion/drake/blob/554c2c2ee400c28feff1ee82ebe3e6ab4e33884e/drake/multibody/test/parser_urdf_test/duplicate_but_same_materials.urdf#L19)):*
> ```XML
>         <cylinder length=".02858" radius="0.001"/>
>       </geometry>
>       <material name="fourbar">
> ```

Rename color "foo_color".

---

*[drake/multibody/test/parser_urdf_test/duplicate_but_same_materials.urdf, line 35 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-KXhh1oUTBO5woeL9PM7:-KXhh1oUTBO5woeL9PM8:b-fwinqq) ([raw file](https://github.com/robotlocomotion/drake/blob/554c2c2ee400c28feff1ee82ebe3e6ab4e33884e/drake/multibody/test/parser_urdf_test/duplicate_but_same_materials.urdf#L35)):*
> ```XML
>         <cylinder length=".05398" radius="0.001"/>
>       </geometry>
>       <material name="fourbar">
> ```

Rename color "foo_color".

---

*[drake/multibody/test/parser_urdf_test/non_conflicting_materials_2.urdf, line 5 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-KXhhGrFOtt1UpyuU4xP:-KXhhGrFOtt1UpyuU4xQ:b5lblu2) ([raw file](https://github.com/robotlocomotion/drake/blob/a912be50fabc3bee748dbf49d9b0ba787ddf7b42/drake/multibody/test/parser_urdf_test/non_conflicting_materials_2.urdf#L5)):*
> ```XML
> <!--
> This URDF contains a robot-level material named "brown" that clashes with a
> link-level material named "brown". When loading this, no exception should be
> ```

Add "The material is defined before the link." before "When".

---

*[drake/multibody/test/parser_urdf_test/non_conflicting_materials_3.urdf, line 5 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-KXhhMf-bLXr2xzGf_FO:-KXhhMf-bLXr2xzGf_FP:bmklupx) ([raw file](https://github.com/robotlocomotion/drake/blob/554c2c2ee400c28feff1ee82ebe3e6ab4e33884e/drake/multibody/test/parser_urdf_test/non_conflicting_materials_3.urdf#L5)):*
> ```XML
> <!--
> This URDF contains a robot-level material named "brown" that clashes with a
> link-level material named "brown". When loading this, no exception should be
> ```

Add "The material is defined after the link." before "When".

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4222)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 5 of 10 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/multibody/test/parser_urdf_test/duplicate_but_same_materials.urdf, line 19 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-KXhh4ytYUctDxOEIykC:-KXjhcjtTOBaUhSV0wUA:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/554c2c2ee400c28feff1ee82ebe3e6ab4e33884e/drake/multibody/test/parser_urdf_test/duplicate_but_same_materials.urdf#L19)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Rename color "foo_color".
</details>
Done.

---

*[drake/multibody/test/parser_urdf_test/duplicate_but_same_materials.urdf, line 35 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-KXhh1oUTBO5woeL9PM7:-KXjhf0sitWsQIt8Z3zy:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/554c2c2ee400c28feff1ee82ebe3e6ab4e33884e/drake/multibody/test/parser_urdf_test/duplicate_but_same_materials.urdf#L35)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Rename color "foo_color".
</details>
Done.

---

*[drake/multibody/test/parser_urdf_test/non_conflicting_materials_2.urdf, line 5 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-KXhhGrFOtt1UpyuU4xP:-KXjhgCSiCp4AqJdx_9l:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a912be50fabc3bee748dbf49d9b0ba787ddf7b42/drake/multibody/test/parser_urdf_test/non_conflicting_materials_2.urdf#L5)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Add "The material is defined before the link." before "When".
</details>
Done.

---

*[drake/multibody/test/parser_urdf_test/non_conflicting_materials_3.urdf, line 5 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-KXhhMf-bLXr2xzGf_FO:-KXjhZXPrvSFhQvMZQWp:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/554c2c2ee400c28feff1ee82ebe3e6ab4e33884e/drake/multibody/test/parser_urdf_test/non_conflicting_materials_3.urdf#L5)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Add "The material is defined after the link." before "When".
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4222)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@drake-jenkins-bot  linux-gcc-ninja-experimental please

david-german-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 5 of 10 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4222#-:-KXkV2Jnj1PvZinP9N-t:bnfp4nl)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Reviewed 1 of 7 files at r1, 1 of 2 files at r6, 1 of 3 files at r7, 3 of 3 files at r8.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4222)*
<!-- Sent from Reviewable.io -->
